src/rgw/driver/sfs/sqlite/sqlite_multipart.cc

```diff
@@ -512,4 +512,71 @@ SQLiteMultipart::remove_multiparts_by_bucket_id_transact(
   return retry.run();
 }
 
+std::optional<DBDeletedMultipartItems>
+SQLiteMultipart::remove_done_or_aborted_multiparts_transact(uint max_items
+) const {
+  DBDeletedMultipartItems ret_parts;
+  auto storage = conn->get_storage();
+  RetrySQLite<DBDeletedMultipartItems> retry([&]() {
+    auto transaction = storage.transaction_guard();
+    // get first the list of parts to be deleted up to max_items
+    ret_parts = storage.select(
+        columns(
+            &DBMultipart::upload_id, &DBMultipart::object_uuid,
+            &DBMultipartPart::id
+        ),
+        inner_join<DBMultipart>(
+            on(is_equal(&DBMultipart::upload_id, &DBMultipartPart::upload_id))
+        ),
+        where(
+            is_equal(&DBMultipart::state, MultipartState::DONE) or
+            is_equal(&DBMultipart::state, MultipartState::ABORTED)
+        ),
+        order_by(&DBMultipartPart::id), limit(max_items)
+    );
+    auto ids = storage.select(
+        &DBMultipartPart::id,
+        inner_join<DBMultipart>(
+            on(is_equal(&DBMultipart::upload_id, &DBMultipartPart::upload_id))
+        ),
+        where(
+            is_equal(&DBMultipart::state, MultipartState::DONE) or
+            is_equal(&DBMultipart::state, MultipartState::ABORTED)
+        ),
+        order_by(&DBMultipartPart::id), limit(max_items)
+    );
+    ceph_assert(ids.size() == ret_parts.size());
+    if (ret_parts.size() == 0) {
+      // nothing to be deleted. We can return now
+      // no need to commit the transaction as nothing was changed
+      return ret_parts;
+    }
+    // remove the parts selected
+    storage.remove_all<DBMultipartPart>(where(in(&DBMultipartPart::id, ids)));
+
+    // now check if the multipart holding the part is empty
```

Comment:
If the MP is DONE or ABORTED new parts are unlikely. Only way I can think of is if something is in flight while doing the state change. With that unlikely how about making this common case fast style? 

Example: 
```
for upload in upload ids:
  try:
   DELETE FROM mulitparts where upload_id in $upload_ids
  except Constraint failed:
   log and move on. next GC run will find this thing again, since there are parts left
```
Another suggestion: build a std::set<upload_id> from get<0> of ret_parts?

Reply:
I did the exception thing with buckets in another PR, but I'm not sure I like it. Looks like forcing things a bit to me.
It's like... I'm doing something that might break a constraint and I know it but I'm catching the exception so nobody sees it. 

I still think we could possibly discuss if we could benefit from "on delete cascade" feature.... that way deletions are very simple, we would just delete parts and sqlite would deal itself with parents with no children.

That way all that "delete things" code would be simpler and cleaner.

Maybe something to talk about in a devs call...

Reply:
> I did the exception thing with buckets in another PR, but I'm not sure I like it. Looks like forcing things a bit to me.
> It's like... I'm doing something that might break a constraint and I know it but I'm catching the exception so nobody sees it.

I think we have EAFP (Easier to ask for forgiveness than permission) vs. LBYL (Look before you leap) here :)

The python glossary explains the two approaches quite well: https://docs.python.org/3/glossary.html#term-EAFP
https://docs.python.org/3/glossary.html#term-LBYL

I'll add it to the devs meeting agenda.

Reply:
> I still think we could possibly discuss if we could benefit from "on delete cascade" feature.... that way deletions are very simple, we would just delete parts and sqlite would deal itself with parents with no children.

If it fits cascading multipart deletes to MP parts children sounds very good. Does the select query at the start of the transaction protect us from a change in parts between select and delete cascade?